### PR TITLE
Add DB clean up BG service, update item service get, post, put

### PR DESCRIPTION
**Pull request by arturascervinskas**
Each time after 'get' is used for a particular key, the expiration period for that key should reset (meaning that only unused values are expiring)
dictionary should have it's internal cleanup mechanism to get rid of "zombie" records and maintain health memory usage by cleaning up the trash. Cleanup interval and max expiration period should be defined in config. User should not be allowed to create records with expiration period longer that the max defined in config.

### CODE REVIEW

File: src/Application/Services/ItemService.cs [hunk 1]
```diff
@@ -39,86 +46,97 @@ public async Task<Item> Get(string key)
...
+        if (itemDto.ExpirationPeriod < MaxItemExPeriod)
+        {
+            if (await _itemRepository.Get(itemDto.Key) is null)
             {
-                Key = itemDto.Key,
-                Value = JsonSerializer.Serialize(itemDto.Value),
-                ExpirationPeriod = ExPeriod,
-                ExpirationDate = DateTime.Now.AddSeconds(ExPeriod)
-            };
+                int exPeriod = itemDto.ExpirationPeriod is 0 ? DefaultExPeriod : itemDto.ExpirationPeriod.Value;
```
Comment (ghost) [inline src/Application/Services/ItemService.cs]: incorrect code, should be "If expiration period is not defined - default value should be used from config"

File: src/Application/Services/ItemService.cs [hunk 2]
```diff
@@ -39,86 +46,97 @@ public async Task<Item> Get(string key)
...
-
-            ItemEntity itemEntity = new()
+            if (await _itemRepository.Get(itemDto.Key) is null)
             {
-                Key = itemDto.Key,
-                Value = JsonSerializer.Serialize(itemDto.Value),
-                ExpirationPeriod = ExPeriod,
-                ExpirationDate = DateTime.Now.AddSeconds(ExPeriod)
-            };
+                int exPeriod = itemDto.ExpirationPeriod is 0 ? DefaultExPeriod : itemDto.ExpirationPeriod.Value;
```
Comment (ghost) [inline src/Application/Services/ItemService.cs]: incorrect code, should be "If expiration period is not defined - default value should be used from config"

File: src/Application/Services/ItemService.cs [hunk 3]
```diff
@@ -39,86 +46,97 @@ public async Task<Item> Get(string key)
...
-            ItemEntity itemEntity = new()
+                return "Item created";
+            }
+            else
             {
-                Key = itemDto.Key,
-                Value = JsonSerializer.Serialize(itemDto.Value),
-                ExpirationPeriod = ExPeriod
-            };
+                int exPeriod = itemDto.ExpirationPeriod is 0 ? DefaultExPeriod : itemDto.ExpirationPeriod.Value;
```
Comment (ghost) [inline src/Application/Services/ItemService.cs]: incorrect code, should be "If expiration period is not defined - default value should be used from config"

File: src/Application/Services/ItemService.cs [hunk 4]
```diff
@@ -2,17 +2,24 @@
...
 using System.Text.Json;
 
 namespace Application.Services;
 
 public class ItemService
 {
     private readonly IItemRepository _itemRepository;
+    private readonly IConfiguration _configuration;
+    public int MaxItemExPeriod { get; }
+    public int DefaultExPeriod { get; }
```
Comment (ghost) [inline src/Application/Services/ItemService.cs]: why it is public, should be private readonly

File: src/Infrastructure/Services/DbCleanupBgService.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                    DateTime date = DateTime.UtcNow;
+
+                    int rowsDeleted = await itemRepository.DeleteExpiredItems(date);
+
+                    _logger.LogInformation("Database cleanup completed successfully. {RowsDeleted} rows deleted.", rowsDeleted);
+                }
+
+                var dbCleanupInterval = _configuration.GetValue<int>("DbCleanupInterval");
+
+                await Task.Delay(TimeSpan.FromSeconds(dbCleanupInterval), stoppingToken);
```
Comment (ghost) [inline src/Infrastructure/Services/DbCleanupBgService.cs]: await should be after try catch, if error could be loop without delay

**Review by ghost — COMMENTED** [review]
code is not compiling

File: src/Infrastructure/Repository/ItemRepository.cs [hunk 6]
```diff
@@ -57,4 +57,19 @@ public async Task Delete(string key)
 
         await _dbConnection.ExecuteAsync(sql, new { key });
     }
+
+    public async Task<int> DeleteExpiredItems(DateTime date)
+    {
+        var sql = @"DELETE FROM Items
+                     WHERE expiration_date <= @ExpiredDate";
```
Comment (ghost) [inline src/Infrastructure/Repository/ItemRepository.cs]: do not use var use string

File: src/Infrastructure/Repository/ItemRepository.cs [hunk 7]
```diff
@@ -57,4 +57,19 @@ public async Task Delete(string key)
...
+
+        return await _dbConnection.ExecuteAsync(sql, new { ExpiredDate = date });
+    }
+
+    public async Task UpdateExDate(string key, DateTime date)
+    {
+        string sql = "UPDATE items SET expiration_date = @ExpirationDate WHERE key = @Key";
+
```
Comment (ghost) [inline src/Infrastructure/Repository/ItemRepository.cs]: inconsistent sql query, use new line like in other

File: src/Infrastructure/Repository/ItemRepository.cs [hunk 8]
```diff
@@ -57,4 +57,19 @@ public async Task Delete(string key)
...
+    public async Task<int> DeleteExpiredItems(DateTime date)
+    {
+        var sql = @"DELETE FROM Items
+                     WHERE expiration_date <= @ExpiredDate";
+
+        return await _dbConnection.ExecuteAsync(sql, new { ExpiredDate = date });
+    }
+
+    public async Task UpdateExDate(string key, DateTime date)
```
Comment (ghost) [inline src/Infrastructure/Repository/ItemRepository.cs]: use entity to pass data

File: src/Infrastructure/Services/DbCleanupBgService.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+                    _logger.LogInformation("Database cleanup completed successfully. {rowsDeleted} rows deleted.", rowsDeleted);
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Error occurred during database cleanup.");
+            }
+            await Task.Delay(TimeSpan.FromSeconds(_dbCleanupInterval), stoppingToken);
+
```
Comment (ghost) [inline src/Infrastructure/Services/DbCleanupBgService.cs:44]: delete empty line

File: src/Infrastructure/Services/DbCleanupBgService.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
```
Comment (ghost) [inline src/Infrastructure/Services/DbCleanupBgService.cs]: delete empty line

File: src/Infrastructure/Services/DbCleanupBgService.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    private readonly IServiceScopeFactory _scopeFactory;
+    private readonly int _dbCleanupInterval;
+
+    public DbCleanupBgService(IConfiguration configuration, ILogger<DbCleanupBgService> logger, IServiceScopeFactory scopeFactory)
+    {
+        _configuration = configuration;
+        _logger = logger;
+        _scopeFactory = scopeFactory;
+        _dbCleanupInterval = _configuration.GetValue<int>("DbCleanupInterval");
+
```
Comment (ghost) [inline src/Infrastructure/Services/DbCleanupBgService.cs]: delete empty line

File: src/Application/Services/ItemService.cs [hunk 12]
```diff
@@ -39,29 +44,47 @@ public async Task<Item> Get(string key)
...
     {
+        if (itemDto.ExpirationPeriod <= 0)
+            throw new ArgumentException("ExpirationPeriod cannot be negative.");
+
+        int exPeriod = itemDto.ExpirationPeriod ?? _defaultExPeriod;
+
+        if (exPeriod > _maxItemExPeriod)
+            throw new ArgumentException($"Invalid ExpirationPeriod field");
+
+
```
Comment (ghost) [inline src/Application/Services/ItemService.cs]: delete empty line

**Review by ghost — APPROVED** [review]

**Review by ViktorasSce — APPROVED** [review]

**Pull request merged**